### PR TITLE
test/{librbd, rgw}: retry when bind fail with port 0

### PR DESCRIPTION
**Pull request by tchaikov**
there is chance that the bind() call may fail if we have another test
happen to pick the free port picked by operating system. in this case,
we just retry up to 42 times with an increasing backoff.

in theory, this change does not fully address the racing, but it should
help to alleviate this issue.

See-also: https://tracker.ceph.com/issues/57116
Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
- https://jenkins.ceph.com/job/ceph-pull-requests/103117/

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
I clicked through all test results links and saw one related failure: https://jenkins.ceph.com/job/ceph-pull-requests/103122.  This is from before a sleep was added, right?

File: src/test/librbd/migration/test_mock_HttpClient.cc [hunk 1]
```diff
@@ -69,12 +69,31 @@ class TestMockMigrationHttpClient : public TestMockFixture {
...
+      } catch (const boost::system::system_error& e) {
+	if (retries == MAX_BIND_RETRIES) {
+	  throw;
+	}
+	if (e.code() != boost::system::errc::address_in_use) {
+	  throw;
+	}
+      }
+      // backoff a little bit
+      sleep(retries);
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/test/librbd/migration/test_mock_HttpClient.cc]: changed and repushed.

File: src/test/librbd/migration/test_mock_HttpClient.cc [hunk 2]
```diff
@@ -69,12 +69,31 @@ class TestMockMigrationHttpClient : public TestMockFixture {
...
+  // if we have a racing where another thread manages to bind and listen the
+  // port picked by this acceptor, try again.
+  static constexpr int MAX_BIND_RETRIES = 42;
+
   void SetUp() override {
     TestMockFixture::SetUp();
 
     ASSERT_EQ(0, open_image(m_image_name, &m_image_ctx));
 
-    create_acceptor(false);
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/test/librbd/migration/test_mock_HttpClient.cc]: yeah. good point. changed accordingly.

**Comment by tchaikov** [comment]
> I clicked through all test results links and saw one related failure: https://jenkins.ceph.com/job/ceph-pull-requests/103122. This is from before a sleep was added, right?

exactly, that's the reason why i added the `sleep()` calls.

File: src/test/rgw/test_http_manager.cc [hunk 3]
```diff
@@ -23,14 +24,43 @@
...
+      } catch (const boost::system::system_error& e) {
+	if (retries == MAX_BIND_RETRIES) {
+	  throw;
+	}
+	if (e.code() != boost::system::errc::address_in_use) {
+	  throw;
+	}
+      }
+      // backoff a little bit
+      sleep(retries * 10'000);
```
Comment (idryomov) [inline src/test/rgw/test_http_manager.cc:53]: I think you missed sleep -> usleep here.

File: src/test/librbd/migration/test_mock_HttpClient.cc [hunk 4]
```diff
@@ -83,10 +84,28 @@ class TestMockMigrationHttpClient : public TestMockFixture {
     TestMockFixture::TearDown();
   }
 
+  // if we have a racing where another thread manages to bind and listen the
+  // port picked by this acceptor, try again.
+  static constexpr int MAX_BIND_RETRIES = 42;
   void create_acceptor(bool reuse) {
```
Comment (idryomov) [inline src/test/librbd/migration/test_mock_HttpClient.cc:90]: Nit: add a newline after `MAX_BIND_RETRIES` declaration.

File: src/test/rgw/test_http_manager.cc (hunk 3 above)
Comment (tchaikov) [reply src/test/rgw/test_http_manager.cc:53]: ahh, right. fixed and repushed.

**Review by cbodley — APPROVED** [review]

**Pull request merged**